### PR TITLE
COMP: Update vendored MetaIO to 2026-09-04 (11606dcd)

### PR DESCRIPTION
**Pull request by hjmjohnson**
Vendored MetaIO update to `11606dcd` (2026-09-04), closing #6835. Brings in the deferred-trailer fix from Kitware/MetaIO#147 and the settable chunk size plus its regression test from Kitware/MetaIO#148.

<details>
<summary>The defect (#6835)</summary>

`MET_PerformUncompression` decompressed in chunks and stopped as soon as the
output buffer was full. When a chunk boundary fell inside the gzip trailer,
the last `inflate` returned `Z_BUF_ERROR` with every output byte already
written and the trailer still sitting in the next unread input piece.

Since #6575 the function requires `Z_STREAM_END` before accepting the data —
the CRC is only validated at stream end — so a perfectly valid file was
rejected. The size of the file determined whether it hit the boundary, which
is why it looked intermittent.

The fix keeps feeding input into a one-byte scratch buffer while the output is
full, so zlib can consume the trailer, verify the CRC, and report
`Z_STREAM_END`. A genuinely truncated stream still has no trailer to find and
is still rejected.

An earlier attempt supplied scratch output without refilling the input. That
cannot work: `avail_in` was already zero, so `inflate` had nothing to consume.
It was caught by a reproduction harness before publishing, not by inspection.

</details>

<details>
<summary>Local verification</summary>

macOS arm64, `ITK_BUILD_DEFAULT_MODULES=OFF` with `Module_ITKIOMeta` and
`Module_ITKTestKernel`:

- Build completes, 0 errors.
- `ctest`: **245/245 pass**.
- `testMeta15UncompressChunkBoundary`, the new upstream regression test, was
  compiled against this vendored copy and passes — including its negative
  case, which confirms a truncated stream is still rejected.

</details>

<details>
<summary>Contents of the update</summary>

| File | Change |
|---|---|
| `src/metaUtils.cxx` | trailer drain loop; `MET_MaxChunkSize` no longer `constexpr` |
| `src/metaUtils.h` | `MET_SetMaxChunkSize` / `MET_GetMaxChunkSize` |
| `src/tests/testMeta15UncompressChunkBoundary.cxx` | new regression test |
| `src/tests/CMakeLists.txt` | registers it |

The settable chunk size exists so the boundary case is reachable from a test
without allocating a gigabyte.

</details>

### CODE REVIEW

**Comment by hjmjohnson** [comment]
MetaIO needs more comprehensive testing in the upstream codebase before considering an update here.

File: Modules/ThirdParty/MetaIO/src/MetaIO/src/metaUtils.cxx [hunk 1]
```diff
@@ -57,7 +57,22 @@ bool META_DEBUG = false;
...
-constexpr static std::streamoff MET_MaxChunkSize = 1024 * 1024 * 1024;
+static std::streamoff MET_MaxChunkSize = 1024 * 1024 * 1024;
+
+void
+MET_SetMaxChunkSize(std::streamoff chunkSize)
+{
+  if (chunkSize > 0)
+  {
+    MET_MaxChunkSize = chunkSize;
+  }
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/ThirdParty/MetaIO/src/MetaIO/src/metaUtils.cxx:68]: Confirmed and fixed upstream in Kitware/MetaIO#149. The cast at the compression `avail_out` does truncate 2^32 to 0, and the setter now ignores anything outside `(0, uInt max]`, with a test asserting an oversized request is rejected.

This file is vendored ThirdParty, so the fix belongs upstream rather than in this PR; ITK will pick it up by re-running `UpdateFromUpstream.sh` once #149 merges.

File: Modules/ThirdParty/MetaIO/src/MetaIO/src/metaUtils.cxx [hunk 2]
```diff
@@ -57,7 +57,22 @@ bool META_DEBUG = false;
...
 
-constexpr static std::streamoff MET_MaxChunkSize = 1024 * 1024 * 1024;
+static std::streamoff MET_MaxChunkSize = 1024 * 1024 * 1024;
+
+void
+MET_SetMaxChunkSize(std::streamoff chunkSize)
+{
+  if (chunkSize > 0)
+  {
+    MET_MaxChunkSize = chunkSize;
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/ThirdParty/MetaIO/src/MetaIO/src/metaUtils.cxx:67]: Half of this is a real bug and is fixed upstream in Kitware/MetaIO#149: the outer decompression loop snapshotted the global into `max_chunk_size` while the inner loop read `MET_MaxChunkSize` directly, so one operation could run with two limits. The inner loop now uses the snapshot.

The remaining process-wide mutability is inherent to the knob, whose purpose is test configuration. Rather than add a mutex, #149 documents the contract in the header: set it before any concurrent codec call. Fixed upstream, not here — vendored ThirdParty.

File: Modules/ThirdParty/MetaIO/src/MetaIO/src/tests/testMeta15UncompressChunkBoundary.cxx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    std::cerr << "FAILED: decompressed content does not match the original\n";
+    return 1;
+  }
+  return 0;
+}
+
+static int
+TestCorruptTrailerStillRejected(const std::vector<unsigned char> & raw, std::vector<unsigned char> compressed)
+{
+  compressed[compressed.size() - 1] ^= 0xFF;
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/ThirdParty/MetaIO/src/MetaIO/src/tests/testMeta15UncompressChunkBoundary.cxx:38]: Correct, and fixed upstream in Kitware/MetaIO#149. The corrupt case ran at the restored default, so the trailer arrived in the first input chunk and the drain loop was never entered. It now runs at both the default and the boundary size.

Verified the new case actually catches a regression: forcing the drain loop to report `Z_STREAM_END` on a zlib error makes the test fail, and reverting makes it pass. Fixed upstream, not here — vendored ThirdParty.

File: Modules/ThirdParty/MetaIO/src/MetaIO/src/metaUtils.cxx [hunk 4]
```diff
@@ -890,6 +905,29 @@ MET_PerformUncompression(const unsigned char * sourceCompressed,
       }
     } while (d_stream.avail_out == 0);
   } while (err != Z_STREAM_END && err >= 0);
+  // The output buffer can fill before the trailer arrives in a later input
+  // chunk; keep feeding input so zlib can reach the CRC and report stream end.
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/ThirdParty/MetaIO/src/MetaIO/src/metaUtils.cxx:909]: Condensed to one line upstream in Kitware/MetaIO#149.

One correction to the rationale, though: the cited requirement is ITK's `AGENTS.md`, and this file is vendored MetaIO, which is not governed by ITK's prose budget. The comment was worth shortening on its own merits, so it is shortened — but ITK style rules do not reach into `Modules/ThirdParty/`, and edits there cannot be made in this PR regardless.

**Comment by hjmjohnson** [comment]
All four review findings are real; all four are fixed upstream in Kitware/MetaIO#149, because this PR only vendors `Modules/ThirdParty/MetaIO/` and ITK does not patch vendored sources in place.

**This PR should merge after #149 lands**, at which point I re-run `UpdateFromUpstream.sh` and the vendored copy carries the fixes. Merging as-is vendors a known latent defect.

<details>
<summary>Disposition of each finding</summary>

| Finding | Disposition |
|---|---|
| P1 — oversized chunk size truncates to 0, non-terminating codec loop | Confirmed. Setter now ignores values outside `(0, uInt max]`. |
| P1 — chunk setting read unsynchronized / twice per operation | Half real: the inner decompression loop read the global instead of the snapshot, so one operation could use two limits. Fixed. The remaining process-wide mutability is documented in the header rather than given a mutex. |
| P2 — corrupt-trailer case never exercised the drain path | Confirmed. Runs at both the default and the boundary chunk size now. |
| P2 — condense the source comment | Done upstream. The cited rule is ITK's `AGENTS.md`, which does not govern vendored MetaIO, but the comment was worth shortening anyway. |

The truncation finding is the one worth reading: `MET_SetMaxChunkSize` validated only `chunkSize > 0`, while the codec casts to zlib's `uInt`, so exactly 4 GiB became `0` — no input consumed, no output capacity, no progress. Only reachable through the setter added in #148, so it is new with that change rather than pre-existing.

Verified the new coverage has teeth by making the drain loop swallow zlib errors: the test fails, and reverting makes it pass.

</details>

**Comment by dzenanz** [comment]
Better open a new PR for the later update.
